template.c

```diff
@@ -0,0 +1,75 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/* Expand template string using environment variables.
+ *
+ * If VAR1 is set to "abc" and VAR2 is not set, template
+ * "var1=${VAR1} var2=${VAR2} end" expands to
+ * "var1=abc var2= end".
+ */
+static void expand_template(char *tmpl, char *result)
+{
+	for (;;) {
+		char *start = strstr (tmpl, "${");
+		if (! start)
+			break;
+
+		char *end = strchr (start, '}');
+		if (! end)
+			break;
+
+		*start = '\0';
+		*end = '\0';
+
+		char const *envvar = getenv (start + 2);
+		if (! envvar)
+			envvar = "";
+
+		strcpy (result, tmpl);
+		result += strlen (tmpl);
+		sprintf (result, "%s", envvar);
+		result += strlen (envvar);
+
+		tmpl = end + 1;
+	}
+
+	/* Copy template part after the last variable. */
+	strcpy (result, tmpl);
+}
+
+int main (int argc, char **argv)
+{
+	char *tmpl = NULL;
+	char buffer[1024];
+	size_t n = 0;
+
+	if (argc != 3)
+	{
+		fputs ("Exactly two arguments expected.\n", stderr);
+		return 1;
+	}
+
+	FILE *in = fopen (argv[1], "rb");
+	if (! in)
+	{
+		perror("Cannot open template");
+		return 2;
+	}
+
+	FILE *out = fopen (argv[2], "wb");
+	if (! out)
+	{
+		perror("Cannot open output file");
+		return 3;
+	}
+
+	while (getline (&tmpl, &n, in) >= 0)
+	{
+		expand_template (tmpl, buffer);
+		fputs (buffer, out);
+	}
+	fclose (out);
```

Comment:
Should in also be closed?

Reply:
Input has beed depleted at this point, so closing it is a bit pointless (unless we care about symetry with output). Output is closed explicitly in order to flush the buffers to the disk.